engine/consensus/matching/engine.go

```diff
@@ -98,1063 +75,169 @@ func New(
-	err = e.clearPools(sealedResultIDs)
-	clearPoolsSpan.Finish()
-	if err != nil {
-		return fmt.Errorf("failed to clean mempools: %w", err)
-	}
-
-	// request execution receipts for unsealed finalized blocks
-	requestReceiptsSpan := e.tracer.StartSpanFromParent(sealingSpan, trace.CONMatchCheckSealingRequestPendingReceipts)
-	err = e.requestPendingReceipts()
-	requestReceiptsSpan.Finish()
-	if err != nil {
-		return fmt.Errorf("could not request pending block results: %w", err)
 	}
-
-	// request result approvals for pending incorporated results
-	requestApprovalsSpan := e.tracer.StartSpanFromParent(sealingSpan, trace.CONMatchCheckSealingRequestPendingApprovals)
-	err = e.requestPendingApprovals()
-	requestApprovalsSpan.Finish()
-	if err != nil {
-		return fmt.Errorf("could not request pending result approvals: %w", err)
-	}
-
-	return nil
 }
 
-// sealableResults returns the IncorporatedResults from the mempool that have
-// collected enough approvals on a per-chunk basis, as defined by the matchChunk
-// function. It also filters out results that have an incorrect sub-graph.
-func (e *Engine) sealableResults() ([]*flow.IncorporatedResult, error) {
-	var results []*flow.IncorporatedResult
-
-	lastFinalized, err := e.state.Final().Head()
-	if err != nil {
-		return nil, fmt.Errorf("failed to get last finalized block: %w", err)
-	}
-
-	// go through the results mempool and check which ones we have collected
-	// enough approvals for
-	for _, incorporatedResult := range e.incorporatedResults.All() {
-		// not finding the block header for an incorporated result is a fatal
-		// implementation bug, as we only add results to the IncorporatedResults
-		// mempool, where _both_ the block that incorporates the result as well
-		// as the block the result pertains to are known
-		block, err := e.headersDB.ByBlockID(incorporatedResult.Result.BlockID)
-		if err != nil {
-			return nil, fmt.Errorf("could not retrieve block: %w", err)
-		}
-
-		// At this point we can be sure that all needed checks on validity of ER
-		// were executed prior to this point, since we perform validation of every ER
-		// before adding it into mempool. Hence, the mempool can contain only valid entries.
+// processPendingEvent saves pending event in corresponding queue for further processing by `Core`.
+// While this function runs in separate goroutine it shouldn't do heavy processing to maintain efficient data polling/pushing.
+func (c *Engine) processPendingEvent(event *Event) {
+	switch event.Msg.(type) {
+	case *flow.ExecutionReceipt:
+		c.engineMetrics.MessageReceived(metrics.EngineMatching, metrics.MessageExecutionReceipt)
 
-		// the chunk assigment is based on the first block in its fork which
-		// contains a receipt that commits to this result.
-		assignment, err := e.assigner.Assign(incorporatedResult.Result, incorporatedResult.IncorporatedBlockID)
-		if state.IsNoValidChildBlockError(err) {
-			continue
+		if c.pendingReceipts.Len() < defaultReceiptQueueCapacity {
```

Comment:
maybe it would be good to log when the queues are full

Reply:
Even better would be to monitor the queue size through a Prometheus metric.